### ISSUE TITLE
Question: ArchiveBox ignoring URLs via docker-compose CLI arguments when not passed via stdin

### ISSUE THREAD

**Issue by stephenmetzger** [ArchiveBox/ArchiveBox#947]
Experiencing some weird issues that I cant fix. 

I'm looking to archive daily weather forecasts from the National Weather Service. Link for my site is https://forecast.weather.gov/MapClick.php?lat=38.2482&lon=-121.6002. 

I am running ArchiveBox in Docker (docker compose), and I can add fine through the WebUI and if I exec into the container and run `archivebox add 'https://forecast.weather.gov/MapClick.php?lat=38.2482&lon=-121.6002'`. But, I'm not able to add using `docker-compose run --rm archivebox add 'https://forecast.weather.gov/MapClick.php?lat=38.2482&lon=-121.6002'`. 

I get the message: `Creating archivebox_archivebox_run ... done`, but don't get the typical: 
```[i] [2022-03-17 04:37:09] ArchiveBox v0.6.2: archivebox add http://google.com
    > /data
[+] [2022-03-17 04:37:10] Adding 1 links to index (crawl depth=0)...
    > Saved verbatim input to sources/1647491830-import.txt
    > Parsed 1 URLs from input (Generic TXT)
    > Found 1 new URLs not already in index

[*] [2022-03-17 04:37:10] Writing 1 links to main index...
    √ ./index.sqlite3

[▶] [2022-03-17 04:37:10] Starting archiving of 1 snapshots in index...

[+] [2022-03-17 04:37:10] "google.com"
    http://google.com
    > ./archive/1647491830.115748
      > title
      > headers
      > pdf
        4 files (689.0 KB) in 0:00:05s
[√] [2022-03-17 04:37:15] Update of 1 pages complete (5.43 sec)
    - 0 links skipped
    - 1 links updated
    - 0 links had errors
    Hint: To manage your archive in a Web UI, run:
        archivebox server 0.0.0.0:8000
``` 
That I expect when a job runs, and nothing is added to the archive folder or webui. 

Also confounding this, is that simple sites (e.g. http://google.com or http://example.com) can be added from the CLI with no issues.

**Comment by pirate** [comment]
What's the output of `docker-compose run --rm archivebox version`?

Also try adding `try: true` and `stdin: true` to your docker-compose file.

**Comment by stephenmetzger** [comment]
version is:
```ArchiveBox v0.6.2
Cpython Linux Linux-5.13.0-35-generic-x86_64-with-glibc2.28 x86_64
IN_DOCKER=True DEBUG=False IS_TTY=True TZ=UTC SEARCH_BACKEND_ENGINE=ripgrep
[i] Dependency versions:
 √  ARCHIVEBOX_BINARY     v0.6.2          valid     /usr/local/bin/archivebox   
 √  PYTHON_BINARY         v3.9.5          valid     /usr/local/bin/python3.9    
 √  DJANGO_BINARY         v3.1.10         valid     /usr/local/lib/python3.9/site-packages/django/bin/django-admin.py
 √  CURL_BINARY           v7.64.0         valid     /usr/bin/curl               
 -  WGET_BINARY           -               disabled  /usr/bin/wget               
 -  NODE_BINARY           -               disabled  /usr/bin/node               
 -  SINGLEFILE_BINARY     -               disabled  /node/node_modules/single-file/cli/single-file
 -  READABILITY_BINARY    -               disabled  /node/node_modules/readability-extractor/readability-extractor
 -  MERCURY_BINARY        -               disabled  /node/node_modules/@postlight/mercury-parser/cli.js
 -  GIT_BINARY            -               disabled  /usr/bin/git                
 -  YOUTUBEDL_BINARY      -               disabled  /usr/local/bin/youtube-dl   
 √  CHROME_BINARY         v90.0.4430.93   valid     /usr/bin/chromium           
 √  RIPGREP_BINARY        v0.10.0         valid     /usr/bin/rg                 
[i] Source-code locations:
 √  PACKAGE_DIR           22 files        valid     /app/archivebox             
 √  TEMPLATES_DIR         3 files         valid     /app/archivebox/templates   
 -  CUSTOM_TEMPLATES_DIR  -               disabled                              
[i] Secrets locations:
 -  CHROME_USER_DATA_DIR  -               disabled                              
 -  COOKIES_FILE          -               disabled                              
[i] Data locations:
 √  OUTPUT_DIR            5 files         valid     /data                       
 √  SOURCES_DIR           56 files        valid     ./sources                   
 √  LOGS_DIR              1 files         valid     ./logs                      
 √  ARCHIVE_DIR           3 files         valid     ./archive                   
 √  CONFIG_FILE           332.0 Bytes     valid     ./ArchiveBox.conf           
 √  SQL_INDEX             272.0 KB        valid     ./index.sqlite3           
```
I tried adding TRY=TRUE and STDIN=TRUE to my docker-compose.yml, but issue persists, not sure if I added the correctly. Should they be environmental variables, like this:

```        environment:
            - ALLOWED_HOSTS=*                   # add any config options you want as env vars
            - MEDIA_MAX_SIZE=750m
            - PUID=1000
            - GUID=1000
            # - SEARCH_BACKEND_ENGINE=sonic     # uncomment these if you enable sonic below
            # - SEARCH_BACKEND_HOST_NAME=sonic
            # - SEARCH_BACKEND_PASSWORD=SecretPassword
            - TRY=TRUE
            - STDIN=TRUE
```

**Comment by stephenmetzger** [comment]
Been trying to figure out how to solve this, but the issue is persisting. Any thoughts?

**Comment by pirate** [comment]
`stdin` and `tty` are `docker-compose` options, not environment variables, add them like so:
```yaml
services:
  archivebox:
    ...
    tty: true
    stdin: true
```

**Comment by stephenmetzger** [comment]
with `stdin` and `tty` added to docker-compose, I get `The Compose file './docker-compose.yml' is invalid because:
Unsupported config option for services.archivebox: 'stdin'`. With only `tty`, issue persists.

**Comment by stephenmetzger** [comment]
I'm trying to install via apt on Ubuntu Server 20.04. The trick is that I want the archive files (just the `/archive` folder) to be saved to a SMB share, and it seems like ArchiveBox wants to also store all of the other files and folders (`/logs`, `/sources`, `Archivebox.conf`, and `index.sqlite`) in the same base directory at the `/archive` folder? 
Did my initial install in the `~/` directory, and added `OUTPUT_DIR = /path_to_smb_share/archivebox/archive` to my `ArchiveBox.conf` file. Threw an error:
```
[X] This folder appears to already have files in it, but no index.sqlite3 present.
    You must run init in a completely empty directory, or an existing data folder.
    Hint: To import an existing data folder make sure to cd into the folder first, then run and run 'archivebox init' to pick up where you left off.
    (Always make sure your data folder is backed up first before updating ArchiveBox)
```
So, I ran `archivebox -init` in the SMB share directory, and got:
```
[+] Initializing a new ArchiveBox v0.6.2 collection...
----------------------------------------------------------------------
[+] Building archive folder structure...
    + ./archive, ./sources, ./logs...
    + ./ArchiveBox.conf...
[+] Building main SQL index and running initial migrations...
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/django/db/backends/base/base.py", line 242, in _commit
    return self.connection.commit()
sqlite3.OperationalError: database is locked
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/django/db/migrations/recorder.py", line 68, in ensure_schema
    editor.create_model(self.Migration)
  File "/usr/local/lib/python3.9/dist-packages/django/db/backends/sqlite3/schema.py", line 36, in __exit__
    super().__exit__(exc_type, exc_value, traceback)
  File "/usr/local/lib/python3.9/dist-packages/django/db/backends/base/schema.py", line 117, in __exit__
    self.atomic.__exit__(exc_type, exc_value, traceback)
  File "/usr/local/lib/python3.9/dist-packages/django/db/transaction.py", line 232, in __exit__
    connection.commit()
  File "/usr/local/lib/python3.9/dist-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/django/db/backends/base/base.py", line 266, in commit
    self._commit()
  File "/usr/local/lib/python3.9/dist-packages/django/db/backends/base/base.py", line 242, in _commit
    return self.connection.commit()
  File "/usr/local/lib/python3.9/dist-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.9/dist-packages/django/db/backends/base/base.py", line 242, in _commit
    return self.connection.commit()
django.db.utils.OperationalError: database is locked
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/bin/archivebox", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/dist-packages/archivebox/cli/__init__.py", line 140, in main
    run_subcommand(
  File "/usr/local/lib/python3.9/dist-packages/archivebox/cli/__init__.py", line 80, in run_subcommand
    module.main(args=subcommand_args, stdin=stdin, pwd=pwd)    # type: ignore
  File "/usr/local/lib/python3.9/dist-packages/archivebox/cli/archivebox_init.py", line 43, in main
    init(
  File "/usr/local/lib/python3.9/dist-packages/archivebox/util.py", line 114, in typechecked_function
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/archivebox/main.py", line 328, in init
    for migration_line in apply_migrations(out_dir):
  File "/usr/local/lib/python3.9/dist-packages/archivebox/util.py", line 114, in typechecked_function
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/archivebox/index/sql.py", line 138, in apply_migrations
    call_command("migrate", interactive=False, stdout=out)
  File "/usr/local/lib/python3.9/dist-packages/django/core/management/__init__.py", line 168, in call_command
    return command.execute(*args, **defaults)
  File "/usr/local/lib/python3.9/dist-packages/django/core/management/base.py", line 371, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.9/dist-packages/django/core/management/base.py", line 85, in wrapped
    res = handle_func(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/django/core/management/commands/migrate.py", line 243, in handle
    post_migrate_state = executor.migrate(
  File "/usr/local/lib/python3.9/dist-packages/django/db/migrations/executor.py", line 91, in migrate
    self.recorder.ensure_schema()
  File "/usr/local/lib/python3.9/dist-packages/django/db/migrations/recorder.py", line 70, in ensure_schema
    raise MigrationSchemaMissing("Unable to create the django_migrations table (%s)" % exc)
django.db.migrations.exceptions.MigrationSchemaMissing: Unable to create the django_migrations table (database is locked)
```

**Comment by pirate** [comment]
The archivebox data dir needs to start as an empty dir in a local drive. ./archive within it can then be synlinked to an SMB share.

```bash
mkdir ~/archivebox
cd ~/archivebox
archivebox init --setup

mv archive /path/to/smb/share/archive
ln -s /path/to/smb/share/archive ./archive
```

**Comment by stephenmetzger** [comment]
Unfortunately, having the same (and more) trouble trying to do this from the `apt` install. I really want to get this working in docker, too, but was willing to try another approach. 

I'm pulling my hair out over this, of course the one link I need to have archived is giving me trouble. Is there any other information I can provide to help you figure out what may be going wrong?

**Comment by pirate** [comment]
Need more info, screenshots, error output, your SMB config, directory layout, etc. anything. Not enough to go on without that.

**Comment by stephenmetzger** [comment]
Thanks, 

I actually ended up figuring out a different way to accomplish what I'm trying to do. Bear in mind I am not a programmer, or developer. Just a relatively new homelabber who wants to automate some work stuff. 

Essentially, I want to use NodeRed to get forecast data from the National Weather Service each morning via their API and (at the same time) trigger ArchiveBox to save the NWS forecast page as a pdf to a SMB share that all employees have access to. If the chance of rain on any day within the 7-day forecast is >=50%, I also want to send an e-mail to a group of employees that contains the path to the NWS archive. 

NodeRed is also running as a docker container, so getting NR to interact with ArchiveBox was a challenge. I was using the BigSSH node to send a command to the host machine, and the `docker-compose run --rm archivebox add ...` was working with pretty much everything I threw at it except for the NWS site. 
So, I found some documentation, and played around, and was able to use 
```
echo "archivebox add 'https://forecast.weather.gov/MapClick.php?lat=38.2482&lon=-121.6002'" | docker exec -i --user archivebox archivebox /bin/bash
```
Things are working for me now. But still have no idea why I was having trouble. If you care to investigate further I'm happy to provide any info you need to troubleshoot.

**Comment by pirate** [comment]
Can you post the full verbatim output of
```bash
docker-compose run --rm archivebox add 'https://forecast.weather.gov/MapClick.php?lat=38.2482&lon=-121.6002'
```

(Archivebox is designed to be used with `docker run` not `docker exec`)

**Comment by stephenmetzger** [comment]
```
root@server02:/data/archivebox# docker-compose run --rm archivebox config --set DEBUG=True
Creating archivebox_archivebox_run ... done
[i] [2022-03-23 22:37:26] ArchiveBox v0.6.2: archivebox config --set DEBUG=True
    > /data

DEBUG=True
root@server02:/data/archivebox# docker-compose run --rm archivebox add --parser=url_list 'https://forecast.weather.gov/MapClick.php?lat=38.2482&lon=-121.6002'
Creating archivebox_archivebox_run ... done
root@server02:/data/archivebox#
```
Screen grab: https://imgur.com/ScQHqbw

**Comment by akhilleusuggo** [comment]
> with `stdin` and `tty` added to docker-compose, I get `The Compose file './docker-compose.yml' is invalid because: Unsupported config option for services.archivebox: 'stdin'`. With only `tty`, issue persists.

Make sure you're using the right compose version. Some version doesn't not support certain options.

**Comment by pirate** [comment]
Try this one instead:
`docker-compose run --rm archivebox add --parser=generic_txt 'https://forecast.weather.gov/MapClick.php?lat=38.2482&lon=-121.6002'`

**Comment by pirate** [comment]
Going to close this since there are a bunch of improvements since it was opened, in particular Docker CLI arg requoting was fixed in a bunch of cases in v0.7.2 that should resolve these comments!

Comment back if ya'll are still having issues and I'll reopen the ticket.
